File: 0046-permutations/0046-permutations.rb
# @param {Integer[]} nums
# @return {Integer[][]}
def permute(nums)
    res = []

    p_helper(nums, 0, [], res, nums.length)

    return res 
end

def p_helper(nums, i, slate, res, n)
    # nums 
    if i == n
        res << slate.dup
        return 
    end 

    for j in (0...nums.length) do 
        slate << nums[j]
        p_helper(nums[0...j] + nums[j+1...nums.length], i+1, slate, res, n)
        slate.pop 
    end 
end 

File: 0046-permutations/0046-permutations.py
class Solution:
    def p_helper(self, nums, i, slate, res, n):
        if i == n:
            res.append(slate[:])
            return 

        for j in range(len(nums)):
            slate.append(nums[j])            
            self.p_helper(nums[0:j] + nums[j+1:], i+1, slate, res, n)
            slate.pop()


    def permute(self, nums: List[int]) -> List[List[int]]:
        res = []

        n = len(nums)

        self.p_helper(nums, 0, [], res, n)

        return res 
        